File: linkedin-profile.php
#!/usr/bin/env php

<?php
require('oauth/OAuth.php');
//require("jsontemplate.php");

$tags = array(
 "{{linkedin-profile}}"
);

class LinkedInProfile {
	private $conf = array();

	private function fix_up($str)
	{
		if (is_string($str)) {
			$str = htmlentities($str, null, FALSE);
			$str = str_replace("\n", "<br/>", $str);
			//$str = str_replace("\n", "<br/>", $str);
			//$str = str_replace("•", "&bull;", $str);
		}
		return $str;
	}

	private function fix_dates(&$arr) {
		$months = array(
			1 => "January",
			2 => "February",
			3 => "March",
			4 => "April",
			5 => "May",
			6 => "June",
			7 => "July",
			8 => "August",
			9 => "September",
			10 => "October",
			11 => "November",
			12 => "December"
		);

		foreach ($arr as $key => $position) {
			$period = '';
			if (isset($position['start-date'])) {
				if (isset($position['start-date']['month'])) {
					$period = $period.$months[$position['start-date']['month']].' ';
				}
				$period = $period.$position['start-date']['year'];
			}
			$period = $period.' &ndash; ';
			if ($position['is-current'] == "true") {
				$period = $period.'Present';
			} else if (isset($position['end-date'])) {
				if (isset($position['end-date']['month'])) {
					$period = $period.$months[$position['end-date']['month']].' ';
				}
				$period = $period.$position['end-date']['year'];
			}
			
			$arr[$key]['period'] = $period;
		}
	}

	private function objectsIntoArray($arrObjData, $arrSkipIndices = array())
	{
		$arrData = array();
		
		// if input is object, convert into array
		if (is_object($arrObjData)) {
			$arrObjData = get_object_vars($arrObjData);
		}
		
		if (is_array($arrObjData)) {
			foreach ($arrObjData as $index => $value) {
			if (is_object($value) || is_array($value)) {
				$value = $this->objectsIntoArray($value, $arrSkipIndices); // recursive call
			}
			if (in_array($index, $arrSkipIndices)) {
				continue;
			}
			$arrData[$index] = $this->fix_up($value);
			}
		}
		return $arrData;
	}

	function __construct() {
		$config = file_get_contents('my-linkedin-config.json');
		$this->conf = json_decode($config, TRUE);
		print "I was called!\n";
	}

	// $url = "https://api.linkedin.com/v1/people/~:public:(first-name,last-name,headline,location,industry,summary,specialties,honors,publications,patents,languages,skills,educations,certifications,picture-url,positions)";

	private function APICall($url) {
		$consumer = new OAuthConsumer($this->conf['consumer_key'], $this->conf['consumer_secret'], NULL);
		$token = new OAuthConsumer($this->conf['oauth_token'], $this->conf['oauth_secret']);
		//three-current-positions or :public: but not both :(
		$req = OAuthRequest::from_consumer_and_token($consumer, $token, "GET", $url, array());
		$req->sign_request(new OAuthSignatureMethod_HMAC_SHA1(), $consumer, $token);
		//print "URL is: ".$acc_req->to_url();

		$data = file_get_contents($req->to_url());
		$xml = new SimpleXMLElement($data);
		$data = $this->objectsIntoArray($xml);
		return $data;
	}

	public function getProfile() {
		$data = $this->APICall("https://api.linkedin.com/v1/people/~:public:(first-name,last-name,headline,location,industry,summary,specialties,honors,publications,patents,languages,skills,educations,certifications,picture-url,positions)");

		$this->fix_dates($data['positions']['position']);
		$this->fix_dates($data['educations']['education']);
		return $data;
	}

	public function getHTML($tag) {
		if ($tag == '{{linkedin-profile}}') {
			$tmpl = file_get_contents('profile.tmpl');
			$data = $this->getProfile();
			$profile_html = JsonTemplateModule::expand($tmpl,$data);
			return $profile_html;
		} else {
			throw new Exception('Unknown tag');
		}
	}
}

register_module($tags, new LinkedInProfile());

?>

File: plugins/linkedin-profile.rb
require 'rubygems'
require 'oauth'
require 'json'
require 'erb'
require 'ostruct'
require 'htmlentities'
require 'helpers/recursive-openstruct.rb'
require 'helpers/plugin.rb'

include ERB::Util

class Profile <RecursiveOpenStruct
  def self.handle_object(object)
    if object.kind_of? String then
      coder = HTMLEntities.new
      coder.encode(object, :named).gsub( "\n", "<br/>" )
    else
      object
    end
  end
end


class LinkedInProfile < Plugin
  def expand
    # Read current configuration
    f = File.open("config/my-linkedin-config.json", "r")
    config = JSON.parse(f.read)
    f.close

      # Check whether the config contains the API key and secret
    if config['consumer_key'] == '' || config['consumer_secret'] == ''
            raise "You need to get a LinkedIn API key and secret first and " \
                "save it in the linkedin-config.json file"
    end

    # If the oauth_token and secret are already set, there's nothing to do
    if config['oauth_token'] == '' && config['oauth_secret'] == ''
            raise "You need to get a LinkedIn oauth token and secret and " \
                "save it in the linkedin-config.json file"
    end

    # Set up LinkedIn specific OAuth API endpoints
    consumer_options = { :site => config['api_host'],
                        :authorize_path => config['authorize_path'],
                        :request_token_path => config['request_token_path'],
                        :access_token_path => config['access_token_path'] }

    consumer = OAuth::Consumer.new(config['consumer_key'], config['consumer_secret'],
                                  consumer_options)

    access_token = OAuth::AccessToken.new(consumer, config['oauth_token'],
                                          config['oauth_secret'])

    # Pick some fields
    fields = ['first-name', 'last-name', 'headline', 'location', 'industry',
              'summary', 'specialties', 'honors', 'publications', 'patents',
              'languages', 'skills', 'educations', 'certifications',
              'picture-url', 'positions'].join(',')
    # Make a request for JSON data"/v1/people/~:(#{fields})"
    json_txt = access_token.get("/v1/people/~:public:(#{fields})",
                                'x-li-format' => 'json').body

    profile = Profile.new_recursive(JSON.parse(json_txt))

    # Open & read template file
    f= File.open("rhtml/profile.rhtml", "r")
    template = f.read
    f.close

    rhtml = ERB.new(template)
    rhtml.result(profile.get_binding)
  end
end
